### DISCUSSION TITLE
Solving exception of `Highs` solver

### DISCUSSION THREAD

**Discussion by Zhanwei-Liu** [metab0t/PyOptInterface#15] [Q&A] [answered]
Hi,

When i use `Highs` to solve [PREP-SHOT](https://github.com/PREP-NexT/PREP-SHOT/tree/PyOptInterface), i found two unexpected conditions. I am not sure whether there are some errors for the interface implementation of `Highs`.

1. `Highs` solving objective value is zero, which is not align with solving results of `Gurobi` and `Mosek`.
2. `Highs` output model to file is significant slower than `Gurobi` and `Mosek`.

Below i show a part of logs for `Highs`, `Gurobi` and `Mosek`, respectively.

+ `Highs`
```bash
Coefficient ranges:
  Matrix [4e-04, 1e+06]
  Cost   [1e+00, 1e+00]
  Bound  [0e+00, 0e+00]
  RHS    [0e+00, 0e+00]
Presolving model
83252 rows, 68936 cols, 228318 nonzeros  0s
42692 rows, 28376 cols, 164746 nonzeros  0s
42670 rows, 28339 cols, 146459 nonzeros  0s
Presolve : Reductions: rows 42670(-395723); columns 28339(-328753); elements 146459(-819124)
Solving the presolved LP
Solving the original LP from the solution after postsolve
Model   status      : Optimal
Objective value     :  0.0000000000e+00
HiGHS run time      :          0.39
```

+ `Gurobi`
```bash
Solved with primal simplex
Iteration    Objective       Primal Inf.    Dual Inf.      Time
  582540    1.0538306e+11   0.000000e+00   0.000000e+00    121s

Solved in 582540 iterations and 121.15 seconds (286.16 work units)
Optimal objective  1.053830552e+11
```
+ `Mosek`

```bash
Primal simplex reoptimization started.
ITER      DEGITER(%)  PFEAS       DFEAS       POBJ                  DOBJ                  TIME
0         0.00        0.00e+00    NA          1.053830552479e+11    NA                    0.01    
162       99.39       0.00e+00    NA          1.053830552479e+11    NA                    0.05    
Primal simplex reoptimization terminated. Time: 0.05    
Simplex reoptimization terminated. Time: 0.07    
Basis identification terminated. Time: 0.69
Optimizer terminated. Time: 9.71    
```
You can reproduce my results use the following scripts. 
```bash
git clone -b PyOptInterface https://github.com/PREP-NexT/PREP-SHOT.git
cd PREP-SHOT
conda env create -f prep-shot.yml
conda activate prep-shot
python run.py
```
You can modify the solving configuration in the `solve_parameters` section of `config.json`. 

BTW, `Mosek` can quickly solve this example than `Gurobi`.
 
Do you have any idea for exceptions above? 

Thank you very much!

**Comment by metab0t** [comment]
I change the `singapore` in config.json to `southeast_asia` and got 0 objective value for all solvers.

As for the output model to file, we simply call `Highs_writeModel` API, so it is the problem of HiGHS.

@Zhanwei-Liu

**Reply by Zhanwei-Liu** [reply]
Sorry for this. I have update `config.json`. Please use `southeast_asia` input to run my model.

**Reply by Zhanwei-Liu** [reply]
Thanks for the explanation!

I set the objective function as follows (`model.py`):
```python
model.obj = poi.ExprBuilder(model.cost)
model.set_objective(model.obj, sense=poi.ObjectiveSense.Minimize)
define_constraints(model, para)
```

I `clone` my project and re-run my model. `Gurobi` and `Mosek` still get correct results instead of zero. You may need to `git pull` to get latest codes.  Thank you!

**Reply by Zhanwei-Liu** [reply]
If you still cannot reproduce my results, please let me know, Thank you very much!

**Comment by metab0t** [comment, marked as answer]
Thank you for spotting a bug in HiGHS interface. I ignored the constant term in linear expression when adding linear constraint.
The fix is pushed at https://github.com/metab0t/PyOptInterface/commit/419cecbaedd31607112003df438529ef98e90605
I will release a new version soon.

**Reply by Zhanwei-Liu** [reply]
Thank you very much!

I have tested the latest version, and the bug is fixed!
